roaringbitmap/src/main/java/org/roaringbitmap/SparseOrderedWriter.java

```diff
@@ -0,0 +1,175 @@
+   * Grab a reference to the underlying bitmap
+   *
+   * @return the underlying bitmap
+   */
+  public RoaringBitmap getUnderlying() {
+    return underlying;
+  }
+
+  /**
+   * Adds the value to the underlying bitmap. The data might
+   * be added to a temporary buffer. You should call "flush"
+   * when you are done.
+   *
+   * @param value the value to add.
+   * @throws IllegalStateException if values are not added in increasing order
+   */
+  @Override
+  public void add(int value) {
+    short key = Util.highbits(value);
+    short low = Util.lowbits(value);
+    if (key != currentKey) {
+      if (Util.compareUnsigned(key, currentKey) < 0) {
+        throw new IllegalStateException("Must write in ascending key order");
+      }
+      flush();
+    }
+
+    int valueIndex = valuesCount++;
+    if (valueIndex > 0 && Util.compareUnsigned(low, values[valueIndex - 1]) <= 0) {
+      throw new IllegalStateException("Must write in ascending order.");
+    }
+    if (valueIndex >= values.length) {
+      valuesCount = values.length;
+      throw new IllegalStateException(
+          String.format("Exceeded limit of %d values per container.", values.length));
+    }
+
+    currentKey = key;
+    values[valueIndex] = low;
+  }
+
+
+  /**
+   * Ensures that any buffered additions are flushed to the underlying bitmap.
+   */
+  @Override
+  public void flush() {
+    if (valuesCount > 0) {
+      RoaringArray highLowContainer = underlying.highLowContainer;
+      // we check that it's safe to append since RoaringArray.append does no validation
+      if (highLowContainer.size > 0) {
+        short key = highLowContainer.getKeyAtIndex(highLowContainer.size - 1);
+        if (Util.compareUnsigned(currentKey, key) <= 0) {
+          throw new IllegalStateException("Cannot write " + currentKey + " after " + key);
+        }
+      }
+      highLowContainer.append(currentKey, buildBestContainer());
+    }
+  }
+
+  @Override
+  public boolean isDirty() {
+    return this.valuesCount > 0;
+  }
+
+  @Override
+  public void clear() {
+    this.valuesCount = 0;
+  }
+
+  private Container buildBestContainer() {
+    Container container;
+    short[] values = Arrays.copyOf(this.values, valuesCount);
+    container = new ArrayContainer(values).runOptimize();
+
+    valuesCount = 0;
+
+    return container;
+  }
+
+  /**
+   * Transfers all non-flushed data contained in this SparseOrderedWriter to a DenseOrderedWriter.
+   * It might receive a DenseOrderedWriter instance in order to populate it.
+   * NOTE: This method clears the current SparseOrderedWriter.
+   *
+   * @param denseWriter a nullable DenseOrderedWriter reference.
+   *                    A new DenseOrderedWriter will be created if this is null.
+   * @return a DenseOrderedWriter instance containing all non-flushed data that was
+   *         present in this SparseOrderedWriter.
+   */
+  public DenseOrderedWriter transfer(DenseOrderedWriter denseWriter) {
+    if (denseWriter == null) {
```

Comment:
I don't follow this method. Shouldn't it just create a new DenseWriter? I don't see a case for reusing an existing instance and once the spill has occurred, I see no evidence that going back to a SparseWriter is helpful.